lib/pbench/server/api/resources/query_apis/__init__.py

```diff
@@ -350,3 +352,185 @@ def _get(self, json_data: JSON, _) -> Response:
+                controller: the controller that generated the dataset
+                name: name of the dataset to publish
+                access: The desired access level of the dataset
+
+            dataset: The associated Dataset object
+        """
+        raise NotImplementedError()
+
+    def complete(self, dataset: Dataset, json_data: JSON, error_count: int) -> None:
+        """
+        Complete a bulk Elasticsearch operation, generally by modifying the
+        source Dataset resource.
+
+        This is an abstract method that must be implemented by a subclass.
+
+        Args:
+
+            dataset: The associated Dataset object
+
+            json_data: JSON dictionary of type-normalized key-value pairs
+                controller: the controller that generated the dataset
+                name: name of the dataset to publish
+                access: The desired access level of the dataset
+
+            error_count: The number of errors encountered during the bulk
+                operation. We want to support idempotency: the Dataset should
+                not be altered unless the error_count is 0.
+        """
+        raise NotImplementedError()
+
+    def _post(self, json_data: JSON, _) -> Response:
+        """
+        Perform the requested POST operation, and handle any exceptions.
+
+        NOTE: This is called by the ApiBase post() method through its dispatch
+        method, which provides parameter validation.
+
+        Args:
+            json_data: Type-normalized client JSON input
+            _: Original incoming Request object (not used)
+
+        Returns:
+            Response to return to client
+        """
+        klasname = self.__class__.__name__
+
+        try:
+            dataset = Dataset.attach(
+                controller=json_data["controller"], name=json_data["name"]
+            )
+        except DatasetNotFound as e:
+            abort(HTTPStatus.NOT_FOUND, message=str(e))
+
+        owner = User.query(id=dataset.owner_id)
+        if not owner:
+            self.logger.error(
+                "Dataset owner ID {} cannot be found in Users", dataset.owner_id
+            )
+            abort(HTTPStatus.INTERNAL_SERVER_ERROR, message="Dataset owner not found")
+
+        # For bulk Elasticsearch operations, we check authorization against the
+        # ownership of a designated dataset rather than having an explicit
+        # "user" JSON parameter. This will raise UnauthorizedAccess on failure.
+        try:
+            self._check_authorization(owner.username, json_data["access"])
+        except UnauthorizedAccess as e:
+            abort(HTTPStatus.FORBIDDEN, message=str(e))
+
+        # Build an Elasticsearch instance to manage the bulk update
+        elastic = elasticsearch.Elasticsearch([self.es_url])
+
+        try:
+            # Pass the bulk command generator to the helper
+            results = elasticsearch.helpers.streaming_bulk(
+                elastic,
```

Comment:
Didn't we use something similar in our py-es-bulk package that implements `streaming_bulk`? Can't we just import and use that package?

Reply:
Check out issue [#19](https://github.com/distributed-system-analysis/py-es-bulk/issues/19) and [#20](https://github.com/distributed-system-analysis/py-es-bulk/issues/20) in the py-es-bulk project.

The problem is that pyesbulk is very narrowly designed for *indexing*, and was hardcoded to expect "create" operation codes... it'll fail if the operation wasn't "create". That won't be too hard to fix (aside from the testing), but the way it handles reporting errors also bothers me because it forces a human-readable output format (pretty-printed JSON) that's harder to consume electronically... I really want to fix that. I don't want to get any of that entangled with an immediate fix for "publish", so I'll handle that separately some other time.